### PR TITLE
feat(@angular-devkit/build-angular): support karma with application builder

### PR DESCRIPTION
**Pull request by jkrems**
## PR Checklist

Please check to confirm your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular-cli/blob/main/CONTRIBUTING.md#-commit-message-guidelines
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

## PR Type

What kind of change does this PR introduce?

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] Other... Please describe:

## What is the current behavior?

`ng test` for projects that are using Karma has a hard dependency on the webpack builder. Apps can't use application builder features in any files loaded in the tests, really.

Issue Number: N/A

## What is the new behavior?

Apps can opt into using the application builder by setting the `builderMode` for the karma builder to either "application" or "detect". The latter only works if they switched the app to the new builder for their development/production builds. Once they do, we will use the new application builder toolchain to build their tests before running them in Karma.

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

## Other information

This PR builds on top of https://github.com/angular/angular-cli/pull/28479. The "meat" is in the [last commit](https://github.com/angular/angular-cli/pull/28416/commits/f71e5c8a652ad4d333d6b34ac3e53de51a249a49).

* High level changes
  - Split non-webpack parts of `find-tests-plugin.ts` into `find-tests.ts` to ensure we use the same list of test files. The code is otherwise unmodified.
  - Split webpack parts of `karma/index.ts` into `karma/browser_builder.ts`. The code is _mostly_ unmodified though moving out the `karmaOptions` initialization changes the order of operations a bit.
  - To handle custom karma config files, I'm currently force-deleting the old framework plugin from the karma config after loading. An alternative would have been to make the presence of that plugin a hard error instead. The idea was to allow trying it out with `--builder-mode=application` without changing a project, even in projects that have a `karma.config` file.
  - For similar reasons I'm currently only warning when `transforms.webpackConfiguration` is present. This could be made a hard error in the future.
* The default remains the "browser" builder, at least for now. It may be viable to make "detect" the default in v19 but that seems risky given that this code hasn't seen a whole lot of real life testing.
* The application builder-based Karma runner supports neither `--watch` nor `--code-coverage`. They are likely both necessary to be a full replacement for many teams. I'm planning to implement at least code coverage in a quick-ish follow-up PR.

### CODE REVIEW

File: packages/angular_devkit/build_angular/src/builders/karma/index.ts [hunk 1]
```diff
@@ -243,22 +152,62 @@ function getBuiltInKarmaConfig(
...
+    const developmentBuilderName = await context.getBuilderNameForTarget(buildTarget);
+
+    return isEsbuildBased(developmentBuilderName);
+  } catch (e) {
+    if (!(e instanceof Error) || e.message !== 'Project target does not exist.') {
+      throw e;
+    }
+    // If we can't find a development builder, we can't use 'detect'.
+    throw new Error(
+      'Failed to detect the detect the builder used by the application. Please set builderMode explicitly.',
```
Comment (clydin) [inline packages/angular_devkit/build_angular/src/builders/karma/index.ts:193]:
```suggestion
      'Failed to detect the builder used by the application. Please set builderMode explicitly.',
```

Comment (jkrems) [reply packages/angular_devkit/build_angular/src/builders/karma/index.ts:193]: Done.

File: packages/angular_devkit/build_angular/src/builders/karma/application_builder.ts [hunk 2]
```diff
@@ -0,0 +1,299 @@
...
+  }
+
+  const testDir = path.join(context.workspaceRoot, 'dist/test-out', randomUUID());
+
+  const [karma, [entryPoints, polyfills]] = await Promise.all([
+    import('karma'),
+    collectEntrypoints(options, context),
+    fs.rm(testDir, { recursive: true, force: true }),
+    // Purge old build disk cache:
+    await purgeStaleBuildCache(context),
```
[the remark this answers is not in the document]
Comment (jkrems) [reply packages/angular_devkit/build_angular/src/builders/karma/application_builder.ts]: Thanks! Updated.

File: packages/angular_devkit/build_angular/src/builders/karma/application_builder.ts [hunk 3]
```diff
@@ -0,0 +1,299 @@
...
+  const tsConfig = await readTsconfig(tsConfigPath);
+
+  const localizePackageInitEntryPoint = '@angular/localize/init';
+  const hasLocalizeType = tsConfig.options.types?.some(
+    (t) => t === '@angular/localize' || t === localizePackageInitEntryPoint,
+  );
+
+  if (hasLocalizeType) {
+    polyfills.push(localizePackageInitEntryPoint);
+  }
```
[the remark this answers is not in the document]
Comment (jkrems) [reply packages/angular_devkit/build_angular/src/builders/karma/application_builder.ts]: Thanks! Updated.

**Comment by jkrems** [comment]
Looks like the test projects hadn't been updated for the polyfill change. Added a commit to explicitly add the localize polyfill to them.

**Pull request merged**